### ISSUE TITLE
Why are TS get/set definedProperties in a class transpiled as enumerable/writable?

### ISSUE THREAD

**Issue by wardbell** [microsoft/TypeScript#3610]
I just confirmed in the TS Playground that the following TS:

```
class Foo {
    get bar() { return true;}
}
```

Produces this JS:

```
var Foo = (function () {
    function Foo() {
    }
    Object.defineProperty(Foo.prototype, "bar", {
        get: function () { return true; },
        enumerable: true,
        configurable: true
    });
    return Foo;
})();
```

This [MDN link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty) confirms that, for `Object.definedProperty` accessors:
• enumerable: Defaults to false.
• configurable: Defaults to false.

Clearly the generated code is overriding the default values for these keys.

Plowing through the ES6 spec is not something I do. So  I’m not sure that ES6 hasn’t changed the defaults. But I’d be surprised.

Now one of the things I miss in TS/ES6 class definitions is the ability to control these keys. It is also not easy to embellish a class with defined properties later … properties that might have non-default definitions

But that’s not my point or my question … which is, why is TS deviating from the defaults? 

I get why I’d want the property to be enumerable most of the time – strange that the spec defaults to false. I don’t understand why I should want it to be configurable (maybe because I don't like something about the property TS generated for me?).

In any case, I don’t get how TS (or ES6) gets away with changing the ES5 defaults.

**Comment by DanielRosenwasser** [comment]
This might actually be a bug.

> <span id="sec-14.5.14" class="secnum">[14.5.14](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-runtime-semantics-classdefinitionevaluation "link to this section")</span> Runtime Semantics: ClassDefinitionEvaluation
=================================================================================================================================================================================================================================

> With parameter className.
>
> &nbsp;&nbsp;&nbsp; *ClassTail* <span class="geq">:</span> *ClassHeritage*<sub>opt</sub> `{` *ClassBody*<sub>opt</sub> `}`
>
> [...]
> For each *ClassElement* *m* in order from *methods*
  1.  If IsStatic of *m* is **false**, then
      1.  Let *status* be the result of performing PropertyDefinitionEvaluation for *m* with arguments *proto* and **false**.
  2.  Else,
      1.  Let *status* be the result of performing PropertyDefinitionEvaluation for *m* with arguments *F* and **false**.

That second parameter to PropertyDefinitionEvaluation is *enumerable*.

Though, PropertyDefinitionEvaluation sets [[Enumerable]] to **true**, so the current behavior may be correct in that respect.

This part of the spec is hard to trace and it would be good to have more eyes on this if possible.

**Comment by gilamran** [comment]
Sorry for polluting this discussion, but why isn't the result code just use ES5 getters/setters ? Why use defineProperty?

**Comment by mhegazy** [comment]
@gilamran you want to put the instance getters/setters on the prototype of the constructor function, and the static ones on the function itself. there is no other way to create something that is both a function, and set properties on it. you can do this with object literals, but then they are not functions.

**Comment by gilamran** [comment]
@mhegazy Got it, thanks!

**Comment by bterlson** [comment]
Sorry I'm late to this thread...

Class methods and getters/setters are created with enumerable false, configurable true, writable true. Seems like an ES6 conformance bug.

(The rationale here was in part to align classes with the built-in types, eg. consider that Array.prototype.* are non-enumerable).

**Comment by DanielRosenwasser** [comment]
Thanks for clarifying @bterlson, and thanks for noticing @wardbell.

**Comment by RyanCavanaugh** [comment]
We should change our emit to align with the ES6 semantics. This will be a breaking change (probably a large one, relatively speaking), but is required for ES5/ES6 compat.

**Comment by calebboyd** [comment]
I believe I've run into this issue. Here is a repeatable snippet that works as intended when targeting es6 or using Babel.
[link](https://babeljs.io/repl/#?evaluate=true&lineWrap=false&presets=es2015%2Creact%2Cstage-2&code=class%20One%20%7B%20%0D%0A%20%20a()%20%7B%7D%0D%0A%7D%0D%0A%0D%0Aclass%20Two%20extends%20One%20%7B%0D%0A%20%20b()%20%7B%7D%0D%0A%20%20c()%20%7B%7D%0D%0A%7D%0D%0A%0D%0Aclass%20Three%20extends%20Two%20%7B%0D%0A%20%20d()%20%7B%7D%0D%0A%20%20e()%20%7B%7D%0D%0A%7D%0D%0A%0D%0Afunction%20uniqueNotConstructor%20(names%2C%20name)%20%7B%0D%0A%20%20~names.indexOf(name)%20%7C%7C%20name%20!%3D%3D%20'constructor'%20%26%26%20names.push(name)%0D%0A%20%20return%20names%0D%0A%7D%0D%0A%0D%0Afunction%20reflectClassProperties%20(source%3A%20Function)%20%7B%0D%0A%20%20const%20properties%3A%20Array%3Cstring%3E%20%3D%20%5B%5D%2C%0D%0A%20%20%20%20constructors%3A%20Array%3CFunction%3E%20%3D%20%5B%5D%0D%0A%0D%0A%20%20let%20current%20%3D%20source%0D%0A%0D%0A%20%20while%20(current%20%26%26%20current.prototype%20%26%26%20current%20!%3D%3D%20Function.prototype)%20%7B%0D%0A%20%20%20%20Object.getOwnPropertyNames(current.prototype)%0D%0A%20%20%20%20%20%20.reduce(uniqueNotConstructor%2C%20properties)%0D%0A%20%20%20%20constructors.push(current)%0D%0A%20%20%20%20current%20%3D%20Object.getPrototypeOf(current)%0D%0A%20%20%7D%0D%0A%0D%0A%20%20return%20%7B%0D%0A%20%20%20%20source%2C%0D%0A%20%20%20%20properties%2C%0D%0A%20%20%20%20constructors%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Aconsole.log(reflectClassProperties(Three))%0D%0A%0D%0A)

**Comment by vvakame** [comment]
Is anyone doing this work?
I met this issue in [SkateJS](http://skate.js.org/).
SkateJS wrote by Babel. SkateJS expected inherit static member of `enumerable: false`.
But TypeScript can't inherit it.

**Comment by andreawyss** [comment]
The more I look at this and the more I think that TypeScript is doing it correctly. 
It Is Babel that must generate class members enumerable = true
One should be able to enumerate through the user defined class members.

**Comment by bloadvenro** [comment]
@wardbell 
Probably I misunderstand, but is it actually a bug?

Since ECMAScript 5.1, when we're working with get/set and object literal style in plain Javascript we have the following:

```
const me = {
    firstName: 'Viktor',
    lastName: 'Petevotov',
    get fullName() {
        return this.firstName + " " + this.lastName
    },
    set fullName(fullName) {
        var _fullName = fullName.split(' ')
        if (_fullName.length === 2) {
            this.firstName = _fullName[0]
            this.lastName = _fullName[1]
        }
    },
}
```

This is how fullName property descriptor looks like when plain Javascript applies get/set:
```
{ get: [Function: get fullName],
  set: [Function: set fullName],
  enumerable: true,
  configurable: true }
```

Look at the end of output, where `enumerable` and `configurable` are set to `true`.

Actually we do not need `Object.defineProperty` to define such stuff as this is also ES5+ feature and this is important, when we talk about how Typescript applies get/set to a class prototype. When TS transpiles such case, we have:

```
var Person = (function () {
    function Person() {
        this._firstName = 'John';
        this._lastName = 'Peterson';
    }
    Object.defineProperty(Person.prototype, "firstName", {
        get: function () {
            return this._firstName;
        },
        set: function (value) {
            this._firstName = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Person.prototype, "lastName", {
        // Read only property.
        get: function () {
            return this._lastName;
        },
        enumerable: true,
        configurable: true
    });
    return Person;
}());
``` 

[But this is the equivalent:](http://es6-features.org/#GetterSetter)

```
var Person = (function () {
    function Person() {
        this._firstName = 'John';
        this._lastName = 'Peterson';
    }
    return Person;
}());
Person.prototype = {
    get firstName () {
      return this._firstName;
    },
    set firstName (value) {
      this._firstName = value;
    },
    // Read only property.
    get lastName() {
      return this._lastName;
    },
}
```

In this case Javascript will do the same stuff as for object literals, and if we look at `Person.prototype.firstName` property descriptor, we will see:

```
{ get: [Function: get],
  set: [Function: set],
  enumerable: true,
  configurable: true }
```

So Typescript does its best to set `enumerable: true` and `configurable: true` attributes just like ES5 does it for get/set by default. `Object.defineProperty` here is just the synonym.

**Comment by Arlen22** [comment]
Any updates on this?

**Comment by coldino** [comment]
I see this is still an issue in 2018.

This is clearly a bug. There is an obvious functional difference in the code generated when targeting es5 and when targeting es6.

I'm working with code that depends on properties being non-enumerable, so TypeScript targeting es5 simply produces broken code. I believe relying on properties being non-enumerable is a reasonable thing to do, especially given this is the behaviour of all current es6 browsers.

Please consider fixing this. If there is too much impact on existing code then an option to control the output variant would be useful and allow a graceful change-over period.

**Comment by michkot** [comment]
This was closed with #32264, but this issue description is ambiguous/misleading, as it makes references to Object.definedProperty defaults and "breaking change" of defaults between ES5 and ES6. I would like to provide clarification for potential future readers.

---

The real problem (as hinted by https://github.com/microsoft/TypeScript/issues/3610#issuecomment-135854603 later in the discussion) was that there was already get()/set(val) defined for object literals at ES5 (with enumerable: true, configurable: true) (see 
https://262.ecma-international.org/5.1/#sec-11.1.5), but when ES6 and classes was introduced, get() and set(val) in ES6 classes got different semantics (**enumerable: false**, but still configurable: true) (see, https://262.ecma-international.org/#sec-method-definitions-runtime-semantics-propertydefinitionevaluation). 
However, TypeScript's get() and set() in classes had the ES5 transpilation semantics taken from object literals, creating a discrepancy between semantics of get()/set(val) in classes when the target was ES5 (object-literal-like created manually using Object.defineProperty) and when it was ES6+ (ES6-native with get/set keywords).

For anyone wondering about the semantics of Object.defineProperty, that itself DID NOT change between ES5 and ES6- see https://www.ecma-international.org/wp-content/uploads/ECMA-262_5th_edition_december_2009.pdf (ES 5, section 8.12.9); 
https://262.ecma-international.org/5.1/#sec-8.12.9 (ES 5.1) also states that all fields of 'Property Descriptor' are by **default false**.

> If a field is absent then its value is considered to be false.

---

I would like to note that as a greenhorn JS/TS programmer, I did not realize that there actually was support for get/set in object literals and that it might have different semantics then get/set in class definitions.

-  TS (at least for v >=1.8 && < 3.9) had the aforementioned "enumerable: true" semantics for class accessors, as described in http://javascript.xgqfrms.xyz/pdfs/TypeScript%20Language%20Specification.pdf TS 1.8, sec 8.7.1, page 134.
- Since there is no updated standard document for TS, the only official documentation is https://www.typescriptlang.org/docs/handbook/classes.html#accessors - but that completely omits the important technical detail of what descriptors will get accessors get - I think it might be nice to add it there so people don't have to comb true JS standard to find it out.
- it might also be nice to warn novice users of "keyof" / TypeScript in general that even though property is part of a type (public in class), it says nothing about the property being enumerable or configurable.

---

Finally, I'd like to add an (opinionated) highlight of important facts from developers PoV when using JS/TS classes, since I had to think about all this when trying to decipher this thread :
- there is (in JS/TS) a "syntactic sugar" for creating accessors in object literals and classes, it will internally boil down to an equivalent of Object.defineProperty with AccessorDescriptor 
- it is inconsistent between object literals and classes
- the syntactic sugar (get/set) lacks expressive power for developer to be able explicitly state what Property Descriptors are to be used (if not counting decorators, that come with a strong warning "!experimental!" https://www.typescriptlang.org/docs/handbook/decorators.html )
- moreover, there is a related change to how class fields are handled in TS https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier, which is again caused by JS going in other direction than TS choose to use in the past; 
- even though [[Define]] semantics for fields (enabled with "useDefineForClassFields" ) makes it possible to enforce "readonly" via explicitly defining the field/property with `writable: false`, TS does not use that opportunity (it would be nice option! - maybe should submit a new issue for that, but it would produce different runtime behavior)

What would be my conclusion for writing classes in JS/TS, taking in account ESNext:
- do not use get/set definitions if you expect to mix object literals and class definitions using them in single codebase 
- do not use get/set definitions when you usually need non-default enumerable (or configurable) "settings" (rather just declare it like a field and use something like https://fettblog.eu/typescript-assertion-signatures/ to safely call Object.defineProperty)
- do not define (ugh, why is it called declaration in the JS proposal, when it's actually a definition?) (public) fields on classes unless you expect the "defineProperty call in constructor"-like behaviour to happen!
- do not define readonly public fields on classes if you want the "non-writability" to be enforced at runtime -> just declare the property and use something like https://fettblog.eu/typescript-assertion-signatures/
- - there is no writeonly typescript keyword or JSDoc annotation, and you can not declare an accessor.. so the only thing you can do is to add field-level comment stating that the property is not writable
